src/analysis/normed_space/inner_product.lean

```diff
@@ -0,0 +1,1524 @@
+/-
+Copyright (c) 2019 Zhouhang Zhou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Zhouhang Zhou, Sébastien Gouëzel, Frédéric Dupuis
+-/
+
+import linear_algebra.bilinear_form
+import linear_algebra.sesquilinear_form
+import analysis.special_functions.pow
+import topology.metric_space.pi_Lp
+import data.complex.is_R_or_C
+
+/-!
+# Inner Product Space
+
+This file defines inner product spaces and proves its basic properties.
+
+An inner product space is a vector space endowed with an inner product. It generalizes the notion of
+dot product in `ℝ^n` and provides the means of defining the length of a vector and the angle between
+two vectors. In particular vectors `x` and `y` are orthogonal if their inner product equals zero.
+
+## Main results
+
+- We define the class `inner_product_space` with a number of basic properties, most
+  notably the Cauchy-Schwarz inequality.
+- We show that if `f i` is an inner product space for each `i`, then so is `Π i, f i`
+- We define `euclidean_space 𝕜 n` to be `n → 𝕜` for any `fintype n`, and show that
+  this an inner product space.
+- Existence of orthogonal projection onto nonempty complete subspace:
+  Let `u` be a point in an inner product space, and let `𝕜` be a nonempty complete subspace.
+  Then there exists a unique `v` in `𝕜` that minimizes the distance `∥u - v∥` to `u`.
+  The point `v` is usually called the orthogonal projection of `u` onto `𝕜`.
+
+## Notation
+
+We locally denote the inner product by `⟪·, ·⟫`. We also provide two notation namespaces:
+`real_inner_product_space` and `complex_inner_product_space`, which introduce the notation
+`⟪·, ·⟫` for the real and the complex inner product respectively.
+
+## Implementation notes
+
+We define both the real and complex cases at the same time using the `is_R_of_C` typeclass.
+We choose the convention that inner products are conjugate linear in the first argument and linear
+in the second.
+
+## TODO
+
+- Fix the section on the existence of minimizers and orthogonal projections to make sure that it
+  also applies in the complex case.
+
+## Tags
+
+inner product space, norm
+
+## References
+*  [Clément & Martin, *The Lax-Milgram Theorem. A detailed proof to be formalized in Coq*]
+*  [Clément & Martin, *A Coq formal proof of the Lax–Milgram theorem*]
+
+The Coq code is available at the following address: <http://www.lri.fr/~sboldo/elfic/index.html>
+-/
+
+noncomputable theory
+
+open is_R_or_C real
+open_locale big_operators classical
+
+variables {F : Type*} {G : Type*}
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜] [normed_algebra ℝ 𝕜] [is_R_or_C 𝕜]
```

Comment:
If I understand correctly, the definition of `[is_R_or_C 𝕜]` requires `[nondiscrete_normed_field 𝕜]`, so is there a reason you need the latter?  (Here and in several other places.)

Reply:
It's not needed for the inner product, but there are natural applications of `is_R_or_C` that would require it, like the Hahn-Banach theorem. For example in `analysis/normed_space/hahn_banach` there is a class called `has_exists_extension_norm_eq` which is effectively a version of `is_R_or_C` specialized for Hahn-Banach (which is pretty much defined as "a field for which Hahn-Banach holds"), and this requires `nondiscrete_normed_field`. Then again, maybe it would only make sense to add this requirement if/when this is ported to `is_R_or_C`.

Reply:
Rather, I was thinking that `is_R_or_C` could have been written to extend `nondiscrete_normed_field`, rather than take it as a parameter.  Of course, I should have asked this at the time of the previous PR.  But did you ever consider that?

Reply:
Ah! No I hadn't really considered it. Are there any advantages to doing that besides making it a bit less cumbersome?

Reply:
It's a common rule, see eg [here](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/Some.20olympiad.20formalisations/near/197868147), and I believe the reason is that it helps typeclass inference go faster.  But the rule is not universal, for example simple yes/no properties (like `complete_space E`) are often kept as separate "mixin" typeclasses.  Arguably `is_R_or_C` is more like a mixin, and anyway it is such an unusual situation that there might be reasons to disregard a standard rule here.  I would ask @sgouezel @urkud or other more expert people.

Reply:
Seems safe to `extend` here. No additional type parameters are being introduced, and given that `is_R_or_C` is "the end of the line", there's no danger involving inheritance diamonds with further extensions.

Reply:
Alright, I'll do that!